src/core/graph.js

```diff
@@ -889,6 +898,55 @@ export class Graph {
     return result;
   }
 
+  /**
+   * Create a new graph, in which some nodes have been contracted together.
+   *
+   * contract takes a list of Contractions, each of which specifies a
+   * replacement node, and a list of old node addresses to map onto the new
+   * node. A new graph will be returned where the new node is added, none of
+   * the old nodes are present, and every edge incident to one of the old nodes
+   * has been re-written so that it is incident to the new node instead.
+   *
+   * If the same node addresses is "old" for several contractions, all incident
+   * edges will be re-written to connect to whichever contraction came last.
+
+   * The original Graph is not mutated.
+   *
+   * contract runs in O(n+e+k), where `n` is the number of nodes, `e` is the
+   * number of edges, and `k` is the number of contractions. If needed, we can
+   * improve the peformance by mutating the original graph instead of creating
+   * a new one.
+   */
+  contract(contractions: $ReadOnlyArray<GraphContraction>): Graph {
+    const remap = new Map();
+    const contracted = new Graph();
+    for (const {old, replacement} of contractions) {
+      for (const addr of old) {
+        remap.set(addr, replacement.address);
+      }
+      contracted.addNode(replacement);
+    }
+    for (const node of this.nodes()) {
+      if (!remap.has(node.address)) {
+        contracted.addNode(node);
+      }
+    }
+    for (const edge of this.edges({showDangling: true})) {
+      let {src, dst} = edge;
+      const srcRemap = remap.get(src);
+      if (srcRemap != null) {
```

Comment:
```js
    for (const edge of this.edges({showDangling: true}) {
      const src = NullUtil.orElse(remap.get(src), edge.src);
      const dst = NullUtil.orElse(remap.get(dst), edge.dst);
      // …
    }
```

Reply:
Right. I originally had that construction, except for a little while I had `remap: Map<NodeAddressT, Node>` which necessitated a different pattern. But now I can go back to that.